cpp/include/raft/neighbors/ivf_flat_helpers.cuh

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <raft/core/resource/cuda_stream.hpp>
+#include <raft/neighbors/detail/ivf_flat_build.cuh>
+#include <raft/neighbors/ivf_flat_types.hpp>
+
+#include <raft/core/device_mdspan.hpp>
+#include <raft/core/resources.hpp>
+
+// #include <omp.h>
+
+namespace raft::neighbors::ivf_flat::helpers {
+/**
+ * @defgroup ivf_flat_helpers Helper functions for manipulationg IVF Flat Index
+ * @{
+ */
+
+namespace codepacker {
+/**
+ * Write one flat code into a block by the given offset. The offset indicates the id of the record in the list. This function interleaves the code and is intended to later copy the interleaved codes over to the IVF list on device.
+ * NB: no memory allocation happens here; the block must fit the record (offset + 1).
+ *
+ * @tparam T
+ *
+ * @param[in] flat_code input flat code
+ * @param[out] block block of memory to write interleaved codes to
+ * @param[in] dim dimension of the flat code
+ * @param[in] veclen size of interleaved data chunks
+ * @param[in] offset how many records to skip before writing the data into the list
+ */
+template <typename T>
+__host__ __device__ void pack_1_interleaved(
```

Comment:
Alright, (un)pack_1 sounds like the better naming convention to avoid confusion. I was also told by @cjnolet that the current IVF-PQ CodePacker needs some refactoring and the names of the IVF-PQ functions can be changed to (un)pack_1 then.

Reply:
IVF-PQ functions can unpack multiple vectors at a time, I would not name them (un)pack_1

Reply:
The way to think of `ivf_flat_codepacker.hpp` is that the codepacker a public API and is the smallest atomic unit (packing and unpacking one single row). This can now be used in host and device functions alike -- for packing and unpacking multiple records either through a kernel, or through a CPU for loop.

Reply:
The helper `pack_full_list` is just an example of using the CodePacker for packing a full list.